### PR TITLE
forkfile: port to using pidfds

### PR DESCRIPTION
**Pull request by brauner**
This ports all of the forkfile() handlers to verify the validity of the target
process via pidfds. This hardens our codebase and specifically means that we
can't accidently operate on a wrong process when under heavy contention.

I'm going to port all subcommand but want to do so step by step to minimize the
risk of subtle breakages. Once they're all switched over I will do another
cleanup sweep.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by brauner** [comment]
@stgraber do we want an API extension for this?

**Comment by brauner** [comment]
Blocked on a Jenkins update.

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
Looks like a legit failure now

**Comment by stgraber** [comment]
@brauner still incomplete?

**Comment by brauner** [comment]
> @brauner still incomplete?

There's still stuff to do but this branch can't be merged anytime. I'll just keep pushing to it and if merged before completely switched over i'll just open another one.

**Comment by brauner** [comment]
Failures expected, as this depends on https://github.com/lxc/go-lxc/pull/137 being merged and Jenkins build cache being refreshed.

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
jenkins: test this please

**Comment by stgraber** [comment]
jenkins: test this please

**Comment by brauner** [comment]
Seems I messed something up while reading. Will fix tomorrow morning.

**Comment by stgraber** [comment]
Things still broken :)

**Comment by brauner** [comment]
On May 20, 2020 11:15:12 PM GMT+02:00, "Stéphane Graber" <notifications@github.com> wrote:
>Things still broken :)

Yeah, I know. I didn't get around to fixing this today. :)
It's also strange because it passes locally.

File: lxd/main_forkfile.go [hunk 1]
```diff
@@ -435,7 +435,7 @@ void forkfile(void)
 	if (pid > 0 || pidfd >= 0) {
 		ns_fd = pidfd_nsfd(pidfd, pid);
 		if (ns_fd < 0)
-		_exit(1);
+			_exit(1);
```
Comment (stgraber) [inline lxd/main_forkfile.go]: What's that?

File: lxd/main_checkfeature.go [hunk 2]
```diff
@@ -330,3 +363,7 @@ func canUseSeccompListener() bool {
 func canUseSeccompListenerContinue() bool {
 	return bool(C.seccomp_notify_aware == 2)
 }
+
+func canUsePidFds() bool {
+	return bool(C.pidfd_aware)
```
Comment (stgraber) [inline lxd/main_checkfeature.go:368]: Shouldn't this also confirm liblxc support?

File: lxd/main_forkfile.go (hunk 1 above)
Comment (brauner) [reply lxd/main_forkfile.go]: Leftover. Wanted to fold that commit.

File: lxd/main_checkfeature.go (hunk 2 above)
Comment (brauner) [reply lxd/main_checkfeature.go:368]: Not really, that's done dynamically. liblxc is smart enough to only allocate a pidfd when it knows the kernel supports it so we can infer liblxc support from InitPidFd() right before we make the call.

Comment (stgraber) [reply lxd/main_checkfeature.go:368]: Aren't we wasting time by always hitting go-lxc for pidfd then either getting told not-supported from that or from liblxc and then hitting the fallback interface?

For other liblxc features we do the detection once at startup and then use that result for all future calls.

Comment (brauner) [reply lxd/main_checkfeature.go:368]: I think we're doing that for seccomp. You need to merge a small fix for liblxc first, though.

**Pull request merged**